design/0003-source-metadata.md

```diff
@@ -0,0 +1,182 @@
+other resource to find a client with the appropriate `dst_X` label.  Not only
+does this make the query awkward, it also means that resource-to-resource
+metrics can only be observed on the client side, never on the server side.  This
+limits our ability to measure network latency.
+
+Adding source metadata to HTTP traffic metrics would enable improvements in the
+Linkerd Grafana dashboard, 3rd party tools that consume Linkerd's Prometheus
+metrics, the controller's StatSummary API, and consequently the `linkerd stat`
+CLI command and Linkerd dashboard.  These improvements are out of the scope of
+this proposal.
+
+# Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+We will use the [Downward
+API](https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#the-downward-api)
+to create and mound a volume in meshed pods which contains a file with the pod
+owner's resource name and kind.  This is very similar to [the approach used to
+populate pod labels on span
+annotations](https://github.com/linkerd/linkerd2/pull/4199).  These values can
+be extracted from pod labels such as `linkerd.io/proxy-deployment`,
+`linkerd.io/proxy-daemonset`, etc.  We can create a simple startup script in the
+proxy container to preprocess these pod labels into the labels that we would
+like the proxy to use.  For example, the Downward API would create a volume with
+a file containing:
+
+```
+app="web-svc"
+linkerd.io/control-plane-ns="linkerd"
+linkerd.io/proxy-deployment="web"
+pod-template-hash="5cb99f85d8"
+```
+
+Our startup script would rewrite this to keep only the labels that we want and
+to put them in the desired format:
+
+```
+resource_name="web"
+resource_kind="deployment"
+```
+
+The proxy would read this file to get a list of labels to apply to its traffic
+metrics.  The way the proxy uses these labels differs for the inbound and
+outbound stacks.
+
+
+For the inbound stack, the proxy will prepend the `dst_` prefix to these labels
+and use them to scope the inbound metrics.  This will result in metrics such as:
+
+```
+request_total{
+  direction="inbound",
+  dst_resource_name="web",
+  dst_resource_kind="deployment",
+  ...
+}
+```
+
+The `dst_` prefix is used like this because on the inbound side, the destination
+is the local resource.  Source labels are also added to the inbound metrics but
+we'll come back to that in a moment.
+
+For the outbound stack, the proxy will prepend the `src_` prefix to the labels
+and use them to scope the outbound metrics.  The corresponding `dst_` labels
+will be populated by the dst metadata from the destination controller.  This
+will result in metrics such as:
+
+```
+request_total{
+  direction="outbound",
+  src_resource_name="web",
+  src_resource_kind="deployment",
+  dst_resource_name="emoji",
+  dst_resource_kind="deployment",
+  ...
+}
+```
+
+The outbound stack will also encode the source labels in an HTTP header of the
```

Comment:
This raises a couple questions:

1. Do we care that this is an HTTP only solution? Seems like there'd be some value in having this for TCP as well.
1. Do we mind modifying the request?
1. What does this look like for inbound, non-meshed traffic?
1. Do care about malicious sources being able to pretend their metrics are from elsewhere?

Reply:
1. This would definitely be useful for TCP as well.  Doing so would require communicating the source metadata on the connection somehow.  I'd call this out-of-scope for now, but interesting future work.
2. We already set some Linkerd specific l5d-* information headers on proxy to proxy requests.
3. Such traffic could simply be missing the `src_X` labels.  Or, if it's preferable, these labels could be set to a blank or placeholder value.  Open to suggestions here.
4. This is a good point.  We should only respect these headers when the source has a validated identity.

Reply:
I'm onboard with most of this doc until we get to this detail. I'd very much prefer to do this at the connection-level: in addition to being able to support TCP metrics, this would substantially reduce the overhead of the feature from a runtime-cost (in terms of latency, cpu) point of view. Because this metadata can only be transferred within the mesh, our HTTP/2 multiplexing further reduces any need to rebroadcast this information.

I think that this RFC should depend on a means to share connection-level context between proxies and not get into those weeds (i.e. let's call it out scope, but blocking implementation). And then I think we ought to get an RFC scoped tightly around that context sharing and how to implement it. Happy to help as needed.